File: sdk/dotNet/Integrations/AzureKeyVault/AzureKeyValueStorage.cs
#nullable enable

using System;
using System.Collections.Generic;
using System.Text.Json;
using System.Threading.Tasks;
using Azure.Identity;
using SecretsManager;
using Azure.Security.KeyVault.Keys.Cryptography;
using Microsoft.Extensions.Logging;
using System.IO;
using System.Text;
using System.Linq;
using Azure.Core;

namespace AzureKeyVault
{

    public class AzureKeyValueStorage : IKeyValueStorage
    {
        private const string DefaultConfigFileLocation = "client-config.json";
        private bool IsInitialized = false;
        private string keyId;
        private CryptographyClient cryptoClient;
        private Dictionary<string, string> config = new();
        private string lastSavedConfigHash;
        private readonly string configFileLocation = DefaultConfigFileLocation;
        private ILogger logger;
        public TokenCredential azureCredentials { get; private set; }

        public AzureKeyValueStorage(string keyId, string? configFileLocation = null, AzureSessionConfig? credentials = null, ILogger? logger = null)
        {
            this.keyId = keyId;
            if (configFileLocation == null)
            {
                configFileLocation = Path.GetFullPath(DefaultConfigFileLocation);
            }
            else
            {
                this.configFileLocation = Path.GetFullPath(configFileLocation);
            }


            // Initialize Azure Key Vault CryptographyClient
            if (credentials != null &&
                !string.IsNullOrEmpty(credentials.TenantId) &&
                !string.IsNullOrEmpty(credentials.ClientId) &&
                !string.IsNullOrEmpty(credentials.ClientSecret))
            {
                // Use ClientSecretCredential when all values are provided
                azureCredentials = new ClientSecretCredential(
                    credentials.TenantId,
                    credentials.ClientId,
                    credentials.ClientSecret);
            }
            else
            {
                // Fallback to DefaultAzureCredential
                azureCredentials = new DefaultAzureCredential();
            }
            cryptoClient = new CryptographyClient(new Uri(keyId), azureCredentials);
            this.logger = GetLogger(logger);
            lastSavedConfigHash = "";
        }

        private async Task InitializeClient(){
            await LoadConfigAsync();
            IsInitialized =  true;
        }

        private ILogger GetLogger(ILogger? logger){
            return logger ?? LoggerFactory.Create(builder =>
            {
                builder.SetMinimumLevel(LogLevel.Information);
                builder.AddConsole();
            }).CreateLogger<AzureKeyValueStorage>();
        }

        public string? GetString(string key)
        {
            if (!IsInitialized){
                InitializeClient().Wait();
            }
            if (config.Count == 0)
            {
                LoadConfigAsync().Wait();
            }
            return config.TryGetValue(key, out var value) ? value : null;
        }

        public void SaveString(string key, string value)
        {
            if (!IsInitialized){
                InitializeClient().Wait();
            }
            if (config.Count == 0)
            {
                LoadConfigAsync().Wait();
            }
            config[key] = value;
            SaveConfigAsync(config).Wait();
        }

        public byte[]? GetBytes(string key)
        {
            if (!IsInitialized){
                InitializeClient().Wait();
            }
            if (config.Count == 0)
            {
                LoadConfigAsync().Wait();
            }

            var stringValue = config.TryGetValue(key, out var result) ? result : null;
            return stringValue == null ? null : CryptoUtils.Base64ToBytes(stringValue);
        }

        public void SaveBytes(string key, byte[] value)
        {
            if (!IsInitialized){
                InitializeClient().Wait();
            }
            if (config.Count == 0)
            {
                LoadConfigAsync().Wait();
            }
            config[key] = CryptoUtils.BytesToBase64(value);
            SaveConfigAsync(config).Wait();
        }

        public void Delete(string key)
        {
            if (!IsInitialized){
                InitializeClient().Wait();
            }
            config.Remove(key);
            SaveConfigAsync(config).Wait();
        }

        public async Task CreateConfigFileIfMissingAsync()
        {
            try
            {
                if (File.Exists(configFileLocation))
                {
                    logger.LogInformation("Config file already exists at: {Path}", configFileLocation);
                    return;
                }

                logger.LogInformation("Config file does not exist at: {Path}", configFileLocation);
                string? directory = Path.GetDirectoryName(configFileLocation);
                if (directory != null && !Directory.Exists(directory))
                {
                    Directory.CreateDirectory(directory);
                }

                // Encrypt an empty configuration and write to the file
                byte[] blob = await IntegrationUtils.EncryptBufferAsync(cryptoClient, "{}", logger);
                await File.WriteAllBytesAsync(configFileLocation, blob);

                logger.LogInformation("Config file created at: {Path}", configFileLocation);
            }
            catch (Exception ex)
            {
                logger.LogError("Error creating config file: {Message}", ex.Message);
            }
        }

        private async Task LoadConfigAsync()
        {
            logger.LogInformation("Loading config file {Path}", configFileLocation);
            await CreateConfigFileIfMissingAsync();
            logger.LogDebug("Created config file in path if missing else validating.. {Path}", configFileLocation);
            // Check if the content is plain JSON
            Dictionary<string, string>? parsedConfig = null;
            Exception? jsonError = null;
            bool decryptionError = false;
            try
            {
                // Read the config file
                byte[] contents;
                try
                {
                    string configData = File.ReadAllText(configFileLocation);

                    parsedConfig = JsonSerializer.Deserialize<Dictionary<string, string>>(configData);
                    contents = Encoding.UTF8.GetBytes(configData);
                    logger.LogDebug("Valid JSON parsed successfully.");
                    if (parsedConfig != null)
                    {
                        config = parsedConfig;
                        await SaveConfigAsync(config);
                        lastSavedConfigHash = ComputeMD5Hash(SerializeConfig(config));
                        return;
                    }
                    logger.LogInformation("Loaded config file {Path}", configFileLocation);
                }
                catch (JsonException ex){
                    logger.LogDebug($"Error parsing valid JSON: {ex.Message}");
                    contents = await File.ReadAllBytesAsync(configFileLocation);
                    jsonError = ex;
                }
                catch (Exception ex)
                {
                    logger.LogError("Failed to load config file {Path}: {Message}", configFileLocation, ex.Message);
                    throw new Exception($"Failed to load config file {configFileLocation}");
                }

                if (contents.Length == 0)
                {
                    logger.LogWarning("Empty config file {Path}", configFileLocation);
                    contents = Encoding.UTF8.GetBytes("{}");
                }

                // If parsing as JSON failed, try decryption
                if (jsonError != null)
                {
                    try
                    {
                        logger.LogDebug("Config file is not a valid JSON file: {Message}", jsonError.Message);
                        
                        string decryptedJson = await IntegrationUtils.DecryptBufferAsync(cryptoClient, contents, logger);
                        parsedConfig = JsonSerializer.Deserialize<Dictionary<string, string>>(decryptedJson);
                        logger.LogDebug("Decrypted config file successfully.");

                        if (parsedConfig != null)
                        {
                            config = parsedConfig;
                            lastSavedConfigHash = ComputeMD5Hash(SerializeConfig(config));
                            return;
                        }
                    }
                    catch (Exception ex)
                    {
                        decryptionError = true;
                        logger.LogDebug("Failed to parse decrypted config file: {Message}", ex.Message);
                        throw new Exception($"Failed to parse decrypted config file {configFileLocation}");
                    }
                }

                if (jsonError != null && decryptionError)
                {
                    logger.LogError("Config file is not a valid JSON file: {Message}", jsonError.Message);
                    throw new Exception($"{configFileLocation} may contain JSON format problems");
                }
            }
            catch (Exception ex)
            {
                logger.LogError("Error loading config: {Message}", ex.Message);
                throw;
            }
        }

        private async Task SaveConfigAsync(Dictionary<string, string>? updatedConfig = null, bool force = false)
        {
            try
            {
                // Retrieve current config
                Dictionary<string, string> currentConfig = config ?? new();
                string configJson = SerializeConfig(currentConfig);
                string configHash = ComputeMD5Hash(configJson);

                // Compare updatedConfig hash with current config hash
                if (updatedConfig != null && updatedConfig.Count > 0)
                {
                    string updatedConfigJson = SerializeConfig(updatedConfig);
                    string updatedConfigHash = ComputeMD5Hash(updatedConfigJson);

                    if (updatedConfigHash != configHash)
                    {
                        configHash = updatedConfigHash;
                        config = new Dictionary<string, string>(updatedConfig);
                    }
                }

                // Check if saving is necessary
                if (!force && configHash == lastSavedConfigHash)
                {
                    logger.LogWarning("Skipped config JSON save. No changes detected.");
                    return;
                }

                // Ensure the config file exists
                await CreateConfigFileIfMissingAsync();

                // Encrypt the config JSON and write to the file
                byte[] blob = await IntegrationUtils.EncryptBufferAsync(cryptoClient, SerializeConfig(config), logger);
                await File.WriteAllBytesAsync(configFileLocation, blob);

                // Update the last saved config hash
                lastSavedConfigHash = configHash;
            }
            catch (Exception ex)
            {
                logger.LogError("Error saving config: {Message}", ex.Message);
            }
        }

        public async Task<string> DecryptConfigAsync(bool autosave = true)
        {
            byte[] ciphertext;
            string plaintext = "";
            if (!IsInitialized){
                InitializeClient().Wait();
            }
            try
            {
                // Read the config file
                if (!File.Exists(configFileLocation))
                {
                    logger.LogError("Config file not found: {File}", configFileLocation);
                    throw new FileNotFoundException($"Config file not found: {configFileLocation}");
                }

                ciphertext = await File.ReadAllBytesAsync(configFileLocation);
                if (ciphertext.Length == 0)
                {
                    logger.LogWarning("Empty config file {File}", configFileLocation);
                    return "";
                }
            }
            catch (Exception ex)
            {
                logger.LogError("Failed to load config file {File}: {Message}", configFileLocation, ex.Message);
                throw new Exception($"Failed to load config file {configFileLocation}");
            }

            try
            {
                // Decrypt the file contents
                plaintext = await IntegrationUtils.DecryptBufferAsync(cryptoClient, ciphertext, logger);
                if (string.IsNullOrWhiteSpace(plaintext))
                {
                    logger.LogInformation("Failed to decrypt config file {File}", configFileLocation);
                }
                else if (autosave)
                {
                    // Optionally autosave the decrypted content
                    await File.WriteAllTextAsync(configFileLocation, plaintext);
                }
            }
            catch (Exception ex)
            {
                logger.LogError("Failed to write decrypted config file {File}: {Message}", configFileLocation, ex.Message);
                throw new Exception($"Failed to write decrypted config file {configFileLocation}");
            }

            return plaintext;
        }

        public async Task<bool> ChangeKeyAsync(string newKeyId)
        {
            if (!IsInitialized){
                InitializeClient().Wait();
            }
            string oldKeyId = keyId;
            CryptographyClient oldCryptoClient = cryptoClient;
            if (!IsInitialized){
                InitializeClient().Wait();
            }
            try
            {
                // Update the key and reinitialize the CryptographyClient
                keyId = newKeyId;
                cryptoClient = new CryptographyClient(new Uri(keyId), azureCredentials);

                await SaveConfigAsync(force: true);
            }
            catch (Exception ex)
            {
                // Restore the previous key and crypto client if the operation fails
                keyId = oldKeyId;
                cryptoClient = oldCryptoClient;

                logger.LogError("Failed to change the key to '{NewKeyId}' for config '{ConfigFile}': {Message}", newKeyId, "config.json", ex.Message);
                throw new Exception($"Failed to change the key for config.json");
            }

            return true;
        }

        private static string SerializeConfig(Dictionary<string, string>? config)
        {
            if (config == null)
            {
                return "{}";
            }
            var sortedKeys = Enumerable.OrderBy(config.Keys, k => k).ToList();
            var sortedConfig = sortedKeys.ToDictionary(k => k, k => config[k]);
            return JsonSerializer.Serialize(sortedConfig, new JsonSerializerOptions { WriteIndented = true });
        }

        private static string ComputeMD5Hash(string input)
        {
            using var md5 = System.Security.Cryptography.MD5.Create();
            byte[] hashBytes = md5.ComputeHash(Encoding.UTF8.GetBytes(input));
            return BitConverter.ToString(hashBytes).Replace("-", "").ToLowerInvariant();
        }

    }

}

File: sdk/javascript/packages/azure/src/AzureKeyValueStorage.ts
import { DefaultAzureCredential, ClientSecretCredential } from "@azure/identity";
import { CryptographyClient } from "@azure/keyvault-keys";

// import { existsSync, mkdirSync, writeFileSync, readFileSync } from 'fs';
import { promises as fs } from 'fs';
import { dirname } from 'path';
import { createHash } from 'crypto';
import { KeyValueStorage, platform } from "@keeper-security/secrets-manager-core";
import { AzureSessionConfig } from "./AzureSessionConfig";
import { decryptBuffer, encryptBuffer } from "./utils";
import { getLogger } from "./Logger";
import { DEFAULT_AZURE_CREDENTIAL_ENVIRONMENTAL_VARIABLE, DEFAULT_JSON_INDENT, DEFAULT_LOG_LEVEL, HEX_DIGEST, MD5_HASH } from "./constants";
import { Logger } from "pino";
import { LoggerLogLevelOptions } from "./enum";


export class AzureKeyValueStorage implements KeyValueStorage {

    private defaultConfigFileLocation: string = "client-config.json";
    private keyId!: string;
    private azureCredentials!: ClientSecretCredential | DefaultAzureCredential;
    private cryptoClient!: CryptographyClient;
    private config!: Record<string, string>;
    private lastSavedConfigHash!: string;
    private logger: Logger;
    private configFileLocation: string;

    getString(key: string): Promise<string | undefined> {
        return this.get(key);
    }
    saveString(key: string, value: string): Promise<void> {
        return this.set(key, value);
    }
    async getBytes(key: string): Promise<Uint8Array | undefined> {
        const bytesString = await this.get(key);
        if (bytesString) {
            return Promise.resolve(platform.base64ToBytes(bytesString));
        }
        return Promise.resolve(undefined);
    }
    saveBytes(key: string, value: Uint8Array): Promise<void> {
        const bytesString = platform.bytesToBase64(value);
        return this.set(key, bytesString);
    }

    getObject?<T>(key: string): Promise<T | undefined> {
        return this.getString(key).then((value) => value ? JSON.parse(value) as T : undefined);
    }
    saveObject?<T>(key: string, value: T): Promise<void> {
        const json = JSON.stringify(value);
        return this.saveString(key, json);
    }

    constructor(keyId: string, configFileLocation: string | null, azSessionConfig: AzureSessionConfig | null, logLevel: LoggerLogLevelOptions | null) {
        /** 
        Initilaizes AzureKeyValueStorage

        key_id URI of the master key - if missing read from env variable which resolves to constant 
                            whose value is given by DEFAULT_AZURE_CREDENTIAL_ENVIRONMENTAL_VARIABLE
        key_id URI may also include version in case key has auto rotate enabled
        ex. key_id = "https://<your vault>.vault.azure.net/keys/<key name>/fe4fdcab688c479a9aa80f01ffeac26"
        The master key needs WrapKey, UnwrapKey privileges

        config_file_location provides custom config file location - if missing read from env KSM_CONFIG_FILE
        az_session_config optional az session config - if missing use default env variables
        https://learn.microsoft.com/en-us/dotnet/api/azure.identity.environmentcredential
        **/
        this.configFileLocation = configFileLocation ?? process.env.KSM_CONFIG_FILE ?? this.defaultConfigFileLocation;
        this.keyId = keyId ?? process.env[DEFAULT_AZURE_CREDENTIAL_ENVIRONMENTAL_VARIABLE];
        this.logger = logLevel == null ? getLogger(DEFAULT_LOG_LEVEL) : getLogger(logLevel);

        if (azSessionConfig) {
            this.logger.debug("validating azure credentials provided and selecting client based on provided credentials");
            const hasAzureSessionConfig = azSessionConfig.tenantId && azSessionConfig.clientId && azSessionConfig.clientSecret;
            if (hasAzureSessionConfig) {
                this.logger.debug("azure credentials provided, selecting ClientSecretCredential");
                this.azureCredentials = new ClientSecretCredential(azSessionConfig.tenantId, azSessionConfig.clientId, azSessionConfig.clientSecret);
            } else {
                this.logger.debug("azure credentials not provided, selecting DefaultAzureCredential");
                this.azureCredentials = new DefaultAzureCredential();
            }
        }
        this.logger.debug("initializing crypto client with key id", this.keyId);
        this.cryptoClient = new CryptographyClient(this.keyId, this.azureCredentials);

        this.lastSavedConfigHash = "";
    }

    async init() {
        await this.loadConfig();
        return this; // Return the instance to allow chaining
    }

    private async loadConfig(): Promise<void> {
        await this.createConfigFileIfMissing();

        try {
            // Read the config file
            let contents: Buffer;
            try {
                contents = await fs.readFile(this.configFileLocation);
                this.logger.info(`Loaded config file ${this.configFileLocation}`);
                // eslint-disable-next-line @typescript-eslint/no-explicit-any
            } catch (err: any) {
                this.logger.error(`Failed to load config file ${this.configFileLocation}: ${err.message}`);
                throw new Error(`Failed to load config file ${this.configFileLocation}`);
            }

            if (contents.length === 0) {
                this.logger.warn(`Empty config file ${this.configFileLocation}, selecting configuration as empty object`);
                contents = Buffer.from("{}");
            }

            // Check if the content is plain JSON
            let config: Record<string, string> | null = null;
            let jsonError;
            let decryptionError = false;
            try {
                const configData = contents.toString();
                config = JSON.parse(configData);
                // Encrypt and save the config if it's plain JSON
                if (config) {
                    this.config = config;
                    await this.saveConfig(config);
                    this.lastSavedConfigHash = createHash(MD5_HASH).update(JSON.stringify(config, Object.keys(config).sort(), DEFAULT_JSON_INDENT)).digest(HEX_DIGEST);
                }
                // eslint-disable-next-line @typescript-eslint/no-explicit-any
            } catch (err: any) {
                jsonError = err;
            }

            if (jsonError) {
                const configJson = await decryptBuffer(this.cryptoClient, contents, this.logger);
                try {
                    config = JSON.parse(configJson);
                    this.config = config ?? {};
                    this.lastSavedConfigHash = createHash(MD5_HASH).update(JSON.stringify(config, Object.keys(this.config).sort(), DEFAULT_JSON_INDENT)).digest(HEX_DIGEST);
                    // eslint-disable-next-line @typescript-eslint/no-explicit-any
                } catch (err: any) {
                    decryptionError = true;
                    this.logger.error(`Failed to parse decrypted config file: ${err.message}`);
                    throw new Error(`Failed to parse decrypted config file ${this.configFileLocation}`);
                }
            }
            if (jsonError && decryptionError) {
                this.logger.info(`Config file is not a valid JSON file: ${jsonError.message}`);
                throw new Error(`${this.configFileLocation} may contain JSON format problems`);
            }
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
        } catch (err: any) {
            this.logger.error(`Error loading config: ${err.message}`);
            throw err;
        }
    }

    private async saveConfig(
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        updatedConfig: Record<string, any> = {},
        force = false
    ): Promise<void> {
        try {
            // Retrieve current config
            const config = this.config || {};
            const configJson = JSON.stringify(config, Object.keys(config).sort(), DEFAULT_JSON_INDENT);
            let configHash = createHash(MD5_HASH).update(configJson).digest(HEX_DIGEST);

            // Compare updatedConfig hash with current config hash
            if (Object.keys(updatedConfig).length > 0) {
                const updatedConfigJson = JSON.stringify(updatedConfig, Object.keys(updatedConfig).sort(), DEFAULT_JSON_INDENT);
                const updatedConfigHash = createHash(MD5_HASH)
                    .update(updatedConfigJson)
                    .digest(HEX_DIGEST);

                if (updatedConfigHash !== configHash) {
                    configHash = updatedConfigHash;
                    this.config = { ...updatedConfig }; // Update the current config
                }
            }

            // Check if saving is necessary
            if (!force && configHash === this.lastSavedConfigHash) {
                this.logger.warn("Skipped config JSON save. No changes detected.");
                return;
            }

            // Ensure the config file exists
            await this.createConfigFileIfMissing();

            // Encrypt the config JSON and write to the file
            const stringifiedConfig = JSON.stringify(this.config, Object.keys(this.config).sort(), DEFAULT_JSON_INDENT);
            const blob = await encryptBuffer(this.cryptoClient, stringifiedConfig, this.logger);
            await fs.writeFile(this.configFileLocation, blob);

            // Update the last saved config hash
            this.lastSavedConfigHash = configHash;
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
        } catch (err: any) {
            this.logger.error("Error saving config:", err.message);
        }
    }

    public async decryptConfig(autosave: boolean): Promise<string> {
        let ciphertext: Buffer;
        let plaintext: string = "";

        try {
            // Read the config file
            ciphertext = await fs.readFile(this.configFileLocation);
            if (ciphertext.length === 0) {
                this.logger.warn(`Empty config file ${this.configFileLocation}`);
                return "";
            }
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
        } catch (err: any) {
            this.logger.error(`Failed to load config file ${this.configFileLocation}: ${err.message}`);
            throw new Error(`Failed to load config file ${this.configFileLocation}`);
        }

        try {
            // Decrypt the file contents
            plaintext = await decryptBuffer(this.cryptoClient, ciphertext, this.logger);
            if (plaintext.length === 0) {
                this.logger.error(`Failed to decrypt config file ${this.configFileLocation}`);
            } else if (autosave) {
                // Optionally autosave the decrypted content
                await fs.writeFile(this.configFileLocation, plaintext);
            }
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
        } catch (err: any) {
            this.logger.error(`Failed to write decrypted config file ${this.configFileLocation}: ${err.message}`);
            throw new Error(`Failed to write decrypted config file ${this.configFileLocation}`);
        }

        return plaintext;
    }

    public async changeKey(newKeyId: string): Promise<boolean> {
        const oldKeyId = this.keyId;
        const oldCryptoClient = this.cryptoClient;

        try {
            // Update the key and reinitialize the CryptographyClient
            this.keyId = newKeyId;
            this.cryptoClient = new CryptographyClient(this.keyId, this.azureCredentials);

            await this.saveConfig({}, true);
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
        } catch (error: any) {
            // Restore the previous key and crypto client if the operation fails
            this.keyId = oldKeyId;
            this.cryptoClient = oldCryptoClient;

            this.logger.error(`Failed to change the key to '${newKeyId}' for config '${this.configFileLocation}': ${error.message}`);
            throw new Error(`Failed to change the key for ${this.configFileLocation}`);
        }
        return true;
    }

    private async createConfigFileIfMissing(): Promise<void> {
        try {
            await fs.access(this.configFileLocation);
            this.logger.info(`Config file already exists at: ${this.configFileLocation.toString()}`);
        } catch {
            this.logger.info(`Config file already exists at: ${this.configFileLocation.toString()}`);
            const dir = dirname(this.configFileLocation);
            try {
                await fs.access(dir);
            } catch {
                await fs.mkdir(dir, { recursive: true });
            }
            // Encrypt an empty configuration and write to the file
            const blob = await encryptBuffer(this.cryptoClient, "{}", this.logger);
            await fs.writeFile(this.configFileLocation, blob);
            this.logger.info(`Config file created at: ${this.configFileLocation.toString()}`);
        }
    }

    public async readStorage(): Promise<Record<string, string>> {
        if (!this.config) {
            await this.loadConfig();
        }
        return Promise.resolve(this.config);
    }

    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    public saveStorage(updatedConfig: Record<string, any>): Promise<void> {
        return this.saveConfig(updatedConfig);
    }

    public async get(key: string): Promise<string> {
        const config = await this.readStorage();
        return Promise.resolve(config[key]);
    }

    public async set(key: string, value: string): Promise<void> {
        const config = await this.readStorage();
        config[key] = value;
        await this.saveStorage(config);
    }

    public async delete(key: string): Promise<void> {
        const config = await this.readStorage();

        if (key in Object.keys(config)) {
            this.logger.debug(`Deleting key ${key} from ${this.configFileLocation}`);
            delete config[key];
        } else {
            this.logger.debug(`Key ${key} not found in ${this.configFileLocation}`);
        }
        await this.saveStorage(config);
    }

    public async deleteAll(): Promise<void> {
        await this.readStorage();
        Object.keys(this.config).forEach(key => delete this.config[key]);
        await this.saveStorage({});
    }

    public async contains(key: string): Promise<boolean> {
        const config = await this.readStorage();
        return Promise.resolve(key in Object.keys(config));
    }

    public async isEmpty(): Promise<boolean> {
        const config = await this.readStorage();
        return Promise.resolve(Object.keys(config).length === 0);
    }
}